File: api-java/samples/server/trinsic/ServiceResultsExample.java
package id.trinsic;

import id.trinsic.api.SessionsApi;
import id.trinsic.serviceresults.ExchangeResultRequest;
import id.trinsic.serviceresults.ResultExchanger;
import id.trinsic.serviceresults.ResultProcessor;
import io.javalin.Javalin;
import io.javalin.http.Context;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Example routes showing how to integrate ServiceResults
 */
public class ServiceResultsExample {
    
    public static void serviceResultsExampleRoutes(Javalin app, SessionsApi session) {
        // Create a ResultExchanger instance
        ResultExchanger exchanger = new ResultExchanger(session);

        // Example route using ServiceResults with enhanced error handling
        app.post("/exchange-result-enhanced", ctx -> {
            try {
                ExchangeResultRequest request = ctx.bodyAsClass(ExchangeResultRequest.class);
                
                // Use ResultExchanger to handle the exchange with built-in validation
                Object result = exchanger.exchangeResultData(request);
                
                // Process the result for additional insights
                var processedResult = ResultProcessor.processSessionResult(result);
                
                // Return both the original result and processed insights
                Map<String, Object> response = new HashMap<>();
                response.put("result", result);
                response.put("insights", processedResult);
                
                ctx.json(response);
            } catch (ResultProcessor.ValidationException ex) {
                ctx.status(400).json(Map.of("error", ex.getMessage()));
            } catch (Exception ex) {
                System.err.println("Failed to exchange result: " + ex.getMessage());
                ctx.status(500).json(Map.of("error", "Internal server error"));
            }
        });

        // Example route for batch processing multiple results
        app.post("/exchange-multiple-results", ctx -> {
            try {
                ExchangeResultRequest[] requests = ctx.bodyAsClass(ExchangeResultRequest[].class);
                
                if (requests == null || requests.length == 0) {
                    ctx.status(400).json(Map.of("error", "Request array is required and cannot be empty"));
                    return;
                }

                List<Object> results = new ArrayList<>();
                List<String> errors = new ArrayList<>();

                for (ExchangeResultRequest request : requests) {
                    try {
                        Object result = exchanger.exchangeResultData(request);
                        results.add(result);
                        errors.add(null);
                    } catch (Exception ex) {
                        results.add(null);
                        errors.add(ex.getMessage());
                    }
                }

                Map<String, Object> response = new HashMap<>();
                response.put("results", results);
                response.put("errors", errors);
                
                ctx.json(response);
            } catch (Exception ex) {
                System.err.println("Failed to exchange multiple results: " + ex.getMessage());
                ctx.status(500).json(Map.of("error", "Internal server error"));
            }
        });

        // Example route showing result validation without exchange
        app.post("/validate-result-request", ctx -> {
            try {
                ExchangeResultRequest request = ctx.bodyAsClass(ExchangeResultRequest.class);
                
                if (request == null) {
                    ctx.status(400).json(Map.of("error", "Request body is required"));
                    return;
                }

                // Just validate the request without exchanging
                ResultProcessor.validateRequest(request);
                
                Map<String, Object> response = new HashMap<>();
                response.put("valid", true);
                response.put("message", "Request is valid");
                
                ctx.json(response);
            } catch (ResultProcessor.ValidationException ex) {
                Map<String, Object> response = new HashMap<>();
                response.put("valid", false);
                response.put("error", ex.getMessage());
                
                ctx.status(400).json(response);
            } catch (Exception ex) {
                System.err.println("Failed to validate request: " + ex.getMessage());
                ctx.status(500).json(Map.of("error", "Internal server error"));
            }
        });

        // Example route showing async processing
        app.post("/exchange-result-async", ctx -> {
            try {
                ExchangeResultRequest request = ctx.bodyAsClass(ExchangeResultRequest.class);
                
                // Use async exchange
                exchanger.exchangeResultDataAsync(request)
                    .thenAccept(result -> {
                        // Process the result for additional insights
                        var processedResult = ResultProcessor.processSessionResult(result);
                        
                        Map<String, Object> response = new HashMap<>();
                        response.put("result", result);
                        response.put("insights", processedResult);
                        
                        ctx.json(response);
                    })
                    .exceptionally(throwable -> {
                        if (throwable.getCause() instanceof ResultProcessor.ValidationException) {
                            ctx.status(400).json(Map.of("error", throwable.getCause().getMessage()));
                        } else {
                            System.err.println("Failed to exchange result async: " + throwable.getMessage());
                            ctx.status(500).json(Map.of("error", "Internal server error"));
                        }
                        return null;
                    });
            } catch (Exception ex) {
                System.err.println("Failed to process async exchange request: " + ex.getMessage());
                ctx.status(500).json(Map.of("error", "Internal server error"));
            }
        });
    }
}

File: api-typescript/samples/server/src/service-results-example.ts
import { NetworkApi, SessionsApi } from "@trinsic/api";
import { Express } from "express";
import { ResultExchanger, ResultProcessor, ValidationError } from "./service-results";

/**
 * Example routes showing how to integrate ServiceResults
 */
export function serviceResultsExampleRoutes(app: Express, networkApi: NetworkApi, sessionsApi: SessionsApi) {
  // Create a ResultExchanger instance
  const exchanger = new ResultExchanger(sessionsApi);

  // Example route using ServiceResults with enhanced error handling
  app.post("/exchange-result-enhanced", async (req: any, res: any) => {
    try {
      const request = req.body;
      
      if (!request) {
        return res.status(400).json({ error: "Request body is required" });
      }

      // Use ResultExchanger to handle the exchange with built-in validation
      const result = await exchanger.exchangeResultData(request);
      
      // Process the result for additional insights
      const processedResult = ResultProcessor.processSessionResult(result);
      
      // Return both the original result and processed insights
      res.json({
        result: result,
        insights: processedResult
      });
    } catch (error) {
      if (error instanceof ValidationError) {
        res.status(400).json({ error: error.message });
      } else {
        console.error("Failed to exchange result:", error);
        res.status(500).json({ error: "Internal server error" });
      }
    }
  });

  // Example route for batch processing multiple results
  app.post("/exchange-multiple-results", async (req: any, res: any) => {
    try {
      const requests = req.body;
      
      if (!Array.isArray(requests) || requests.length === 0) {
        return res.status(400).json({ error: "Request array is required and cannot be empty" });
      }

      const results = await exchanger.exchangeMultipleResultsSettled(requests);
      
      const processedResults = results.map(result => {
        if (result.status === 'fulfilled') {
          return result.value;
        } else {
          return null;
        }
      });

      const errors = results.map(result => {
        if (result.status === 'rejected') {
          return result.reason.message || result.reason;
        } else {
          return null;
        }
      });

      res.json({
        results: processedResults,
        errors: errors
      });
    } catch (error) {
      console.error("Failed to exchange multiple results:", error);
      res.status(500).json({ error: "Internal server error" });
    }
  });

  // Example route showing result validation without exchange
  app.post("/validate-result-request", async (req: any, res: any) => {
    try {
      const request = req.body;
      
      if (!request) {
        return res.status(400).json({ error: "Request body is required" });
      }

      // Just validate the request without exchanging
      ResultProcessor.validateRequest(request);
      
      res.json({
        valid: true,
        message: "Request is valid"
      });
    } catch (error) {
      if (error instanceof ValidationError) {
        res.status(400).json({
          valid: false,
          error: error.message
        });
      } else {
        console.error("Failed to validate request:", error);
        res.status(500).json({ error: "Internal server error" });
      }
    }
  });

  // Example route showing batch validation
  app.post("/validate-multiple-requests", async (req: any, res: any) => {
    try {
      const requests = req.body;
      
      if (!Array.isArray(requests)) {
        return res.status(400).json({ error: "Request array is required" });
      }

      const { valid, invalid } = ResultProcessor.validateMultipleRequests(requests);
      
      res.json({
        validCount: valid.length,
        invalidCount: invalid.length,
        valid: valid.map(r => ResultProcessor.sanitizeForLogging(r)),
        invalid: invalid
      });
    } catch (error) {
      console.error("Failed to validate multiple requests:", error);
      res.status(500).json({ error: "Internal server error" });
    }
  });
}